crates/ruff_server/src/edit/notebook.rs

```diff
@@ -0,0 +1,189 @@
+use std::{collections::HashMap, hash::BuildHasherDefault};
+
+use anyhow::Ok;
+use lsp_types::{NotebookCellKind, Url};
+use rustc_hash::FxHashMap;
+
+use crate::{PositionEncoding, TextDocument};
+
+use super::DocumentVersion;
+
+/// The state of a notebook document in the server. Contains an array of cells whose
+/// contents are internally represented by [`TextDocument`]s.
+#[derive(Clone, Debug)]
+pub(crate) struct NotebookDocument {
+    cells: Vec<NotebookCell>,
+    metadata: ruff_notebook::RawNotebookMetadata,
+    version: DocumentVersion,
+    // Used to quickly find the index of a cell for a given URL.
+    cell_index: FxHashMap<lsp_types::Url, usize>,
+}
+
+#[derive(Clone, Debug)]
+struct NotebookCell {
+    url: Url,
+    kind: NotebookCellKind,
+    document: TextDocument,
+}
```

Comment:
I think this might be ok but an alternate way to model this would be to only store the URI of the text document in `document` field. The actual `TextDocument` would be stored centrally where the other `TextDocument` (representing the Python file) is being stored. This would introduce an indirection if we want to get the cell content but it would keep the representation similar to the spec.

Reply:
This would add complexity to document queries, since we can't just fetch documents from the index dynamically in a background thread, so we'd have to fetch all cell text documents in advance and store them in a separate index in the query. I'm OK with our internal representation not matching the LSP specification if it makes things simpler.